### PR TITLE
move model for export preparation to be able reuse in other exporters

### PR DESCRIPTION
**Pull request by eaidova**
# What does this PR do?

This PR propose to move model preparation to export outside onnx exporter for ability reuse it in different exporters (e.g. openvino, neuron) that may have own export configs. Changes are backward compatible for onnx export, should not affect other components, but will help in reusage models preparation (partitioning, building export configs) in other optimum exporters

### CODE REVIEW

File: optimum/exporters/model_preparation.py [hunk 1]
```diff
@@ -0,0 +1,562 @@
+# coding=utf-8
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
```
Comment (echarlaix) [inline optimum/exporters/model_preparation.py:9]: would rename it to utils.py

Comment (eaidova) [reply optimum/exporters/model_preparation.py]: renamed

File: optimum/exporters/utils.py [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+        models_for_export["model"] = (models_for_export["model"], export_config)
+
+    return models_for_export
+
+
+def get_stable_diffusion_models_for_export(
+    pipeline: "StableDiffusionPipeline",
+    int_dtype: str = "int64",
+    float_dtype: str = "fp32",
+    exporter_type: str = "onnx",
```
Comment (echarlaix) [inline optimum/exporters/utils.py:265]: to keep consistency 
```suggestion
    exporter: str = "onnx",
```

**Review by echarlaix — APPROVED** [review]
LGTM !

File: optimum/exporters/tasks.py [hunk 3]
```diff
@@ -1121,10 +1121,11 @@ def decorator(config_cls: Type) -> Type:
                 mapping = supported_model_type_for_library.get(model_type, {})
                 mapping_backend = mapping.get(backend, {})
                 for task in supported_tasks:
-                    if task not in cls.get_all_tasks():
+                    normalized_task = task.replace("-with-past", "")
+                    if normalized_task not in cls.get_all_tasks():
```
Comment (fxmarty) [inline optimum/exporters/tasks.py:1125]: Is this change related to the refactor?

Comment (fxmarty) [inline optimum/exporters/tasks.py:1125]: Is this change related to the refactor?

**Review by fxmarty — COMMENTED** [review]
It looks OK to me, but keep in mind optimum main repo CI is only testing ONNX export.

File: optimum/exporters/tasks.py (hunk 3 above)
Comment (eaidova) [reply optimum/exporters/tasks.py:1125]: it is not related to refactoring, it fixes bug that prevent registering tasks with-past in this case (I found it testing on optimum-intel side for registering configs for openvino)

**Comment by eaidova** [comment]
> @eaidova Could you push an empty commit (or alternatively merge main), so that we run with #1707 & run the CI?

@fxmarty done

**Comment by fxmarty** [comment]
@eaidova Thank you, can you install `ruff==0.1.5` and run `make style`?

**Comment by fxmarty** [comment]
yes please, sorry if it is unrelated

**Review by fxmarty — APPROVED** [review]

**Comment by fxmarty** [comment]
Can you fix
```
exporters/onnx/test_onnx_export.py:30: in <module>
    from optimum.exporters.onnx import (
E   ImportError: cannot import name 'get_decoder_models_for_export' from 'optimum.exporters.onnx' (/home/runner/work/optimum/optimum/optimum/exporters/onnx/__init__.py)
```

Maybe it's good to keep backward compatibility in `exporters/onnx/utils.py` (just importing from `exporters/utils.py` & redefining the functions), with warnings that the functions will be removed in a future version.

**Comment by eaidova** [comment]
> Can you fix
> 
> ```
> exporters/onnx/test_onnx_export.py:30: in <module>
>     from optimum.exporters.onnx import (
> E   ImportError: cannot import name 'get_decoder_models_for_export' from 'optimum.exporters.onnx' (/home/runner/work/optimum/optimum/optimum/exporters/onnx/__init__.py)
> ```
> 
> Maybe it's good to keep backward compatibility in `exporters/onnx/utils.py` (just importing from `exporters/utils.py` & redefining the functions), with warnings that the functions will be removed in a future version. Although not explicitly exposed, maybe some people use these utils.

@fxmarty  added

File: optimum/exporters/onnx/utils.py [hunk 4]
```diff
@@ -190,3 +212,50 @@ def _get_submodels_and_onnx_configs(
...
+    float_dtype: str = "fp32",
+) -> Dict[str, Tuple[Union["PreTrainedModel", "ModelMixin"], "ExportConfig"]]:
+    logger.warn(
+        "usage `optimum.exporters.onnx.utils.get_stable_diffusion_models_for_export` deprecated, please use `optimum.exporters.utils.get_stable_diffusion_models_for_export` instead"
+    )
+    return _get_stable_diffusion_models_for_export(pipeline, int_dtype, float_dtype, exporter="onnx")
+
+
+def get_sam_models_for_export(model: Union["PreTrainedModel", "TFPreTrainedModel"], config: "ExportConfig"):
+    logger.warn(
```
Comment (fxmarty) [inline optimum/exporters/onnx/utils.py:229]: `warn` is deprecated, can you change to `warning`?

File: optimum/exporters/onnx/utils.py [hunk 5]
```diff
@@ -190,3 +212,50 @@ def _get_submodels_and_onnx_configs(
...
+        "usage `optimum.exporters.onnx.utils.get_sam_models_for_export` deprecated, please use `optimum.exporters.utils.get_sam_models_for_export` instead"
```
Comment (fxmarty) [inline optimum/exporters/onnx/utils.py:230]:
```suggestion
        "The usage of `optimum.exporters.onnx.utils.get_sam_models_for_export` is deprecated and will be removed in a future release, please use `optimum.exporters.utils.get_sam_models_for_export` instead."
```

File: optimum/exporters/onnx/utils.py (hunk 4 above)
Comment (eaidova) [reply optimum/exporters/onnx/utils.py]: updated

**Comment by fxmarty** [comment]
Thank you LGTM, failing tests are unrelated!
